src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -2236,60 +2237,53 @@ public Type visitErrorType(ErrorType t, Symbol sym) {
         };
 
     /**
-     * Return the base type of t or any of its outer types that starts
-     * with the given symbol.  If none exists, return null.
+     * Return the base type of t or any of its outer types (by following the types enclosing type)
+     * that start with the given symbol.
+     * If none exists, return null.
      *
      * @param t a type
      * @param sym a symbol
      */
     public Type asOuterSuper(Type t, Symbol sym) {
-        switch (t.getTag()) {
-        case CLASS:
-            do {
-                Type s = asSuper(t, sym);
-                if (s != null) return s;
-                t = t.getEnclosingType();
-            } while (t.hasTag(CLASS));
-            return null;
-        case ARRAY:
-            return isSubtype(t, sym.type) ? sym.type : null;
-        case TYPEVAR:
-            return asSuper(t, sym);
-        case ERROR:
-            return t;
-        default:
-            return null;
-        }
+        return asOuter(t, sym, Type::getEnclosingType);
     }
 
     /**
-     * Return the base type of t or any of its enclosing types that
-     * starts with the given symbol.  If none exists, return null.
+     * Return the base type of t or any of its outer types (by following owner's
+     * enclosing class) that start with the given symbol.
+     * If none exists, return null.
      *
      * @param t a type
      * @param sym a symbol
      */
     public Type asEnclosingSuper(Type t, Symbol sym) {
-        switch (t.getTag()) {
-        case CLASS:
-            do {
-                Type s = asSuper(t, sym);
-                if (s != null) return s;
-                Type outer = t.getEnclosingType();
-                t = (outer.hasTag(CLASS)) ? outer :
-                    (t.tsym.owner.enclClass() != null) ? t.tsym.owner.enclClass().type :
-                    Type.noType;
-            } while (t.hasTag(CLASS));
-            return null;
-        case ARRAY:
-            return isSubtype(t, sym.type) ? sym.type : null;
-        case TYPEVAR:
-            return asSuper(t, sym);
-        case ERROR:
-            return t;
-        default:
-            return null;
+        return asOuter(t, sym, type -> getOwnerEnclosingClassType(type));
```

Comment:
You could also use a method reference here.

Also... I guess here we have to decide whether it's better to use a method reference, and leave a method like `getOwnerEnclosingClassType` permanently there, or, perhaps, just accept the fact that the logic represented by `getOwnerEnclosingClassType` is a bit weird, and has only really one use, and it belongs to a lambda expression, so as to achieve full encapsulation.

Reply:
Also, the one is on `Type::getEnclosingType` and the other would be on `Types::getOwnerEnclosingClassType`.

I this this is too special to make it part of the hierarchy of `Type` and specially the `class DeclaredType`.
I think I would slightly prefer to keep it as a lambda ultimately to promote the custom logic here. I would strongly prefer to inline it even in the lambda, but there we lose the nice name.

Reply:
I agree that as a method it would look confusing either in `Type` or in `Symbol`